halo2_middleware/src/zal.rs

```diff
@@ -0,0 +1,309 @@
+    }
+}
+
+// ZAL using Halo2curves as a backend
+// ---------------------------------------------------
+
+pub mod impls {
+    use std::marker::PhantomData;
+
+    use crate::zal::traits::MsmAccel;
+    use halo2curves::msm::best_multiexp;
+    use halo2curves::CurveAffine;
+
+    // Halo2curve Backend
+    // ---------------------------------------------------
+    #[derive(Default)]
+    pub struct H2cEngine;
+
+    pub struct H2cMsmCoeffsDesc<'c, C: CurveAffine> {
+        raw: &'c [C::Scalar],
+    }
+
+    pub struct H2cMsmBaseDesc<'b, C: CurveAffine> {
+        raw: &'b [C],
+    }
+
+    impl H2cEngine {
+        pub fn new() -> Self {
+            Self {}
+        }
+    }
+
+    impl<C: CurveAffine> MsmAccel<C> for H2cEngine {
+        fn msm(&self, coeffs: &[C::Scalar], bases: &[C]) -> C::Curve {
+            best_multiexp(coeffs, bases)
+        }
+
+        // Caching API
+        // -------------------------------------------------
+
+        type CoeffsDescriptor<'c> = H2cMsmCoeffsDesc<'c, C>;
+        type BaseDescriptor<'b> = H2cMsmBaseDesc<'b, C>;
+
+        fn get_coeffs_descriptor<'c>(&self, coeffs: &'c [C::Scalar]) -> Self::CoeffsDescriptor<'c> {
+            // Do expensive device/library specific preprocessing here
+            Self::CoeffsDescriptor { raw: coeffs }
+        }
+        fn get_base_descriptor<'b>(&self, base: &'b [C]) -> Self::BaseDescriptor<'b> {
+            Self::BaseDescriptor { raw: base }
+        }
+
+        fn msm_with_cached_scalars(
+            &self,
+            coeffs: &Self::CoeffsDescriptor<'_>,
+            base: &[C],
+        ) -> C::Curve {
+            best_multiexp(coeffs.raw, base)
+        }
+
+        fn msm_with_cached_base(
+            &self,
+            coeffs: &[C::Scalar],
+            base: &Self::BaseDescriptor<'_>,
+        ) -> C::Curve {
+            best_multiexp(coeffs, base.raw)
+        }
+
+        fn msm_with_cached_inputs(
+            &self,
+            coeffs: &Self::CoeffsDescriptor<'_>,
+            base: &Self::BaseDescriptor<'_>,
+        ) -> C::Curve {
+            best_multiexp(coeffs.raw, base.raw)
+        }
+    }
+
+    // Backend-agnostic engine objects
+    // ---------------------------------------------------
+    #[derive(Debug)]
+    pub struct PlonkEngine<C: CurveAffine, MsmEngine: MsmAccel<C>> {
+        pub msm_backend: MsmEngine,
+        _marker: PhantomData<C>, // compiler complains about unused C otherwise
+    }
+
+    #[derive(Default)]
+    pub struct PlonkEngineConfig<C, M> {
+        curve: PhantomData<C>,
+        msm_backend: Option<M>,
+    }
+
+    #[derive(Default)]
+    pub struct NoCurve;
+
+    #[derive(Default)]
+    pub struct HasCurve<C: CurveAffine>(PhantomData<C>);
+
+    #[derive(Default)]
+    pub struct NoMsmEngine;
+
+    pub struct HasMsmEngine<C: CurveAffine, M: MsmAccel<C>>(M, PhantomData<C>);
+
+    impl PlonkEngineConfig<NoCurve, NoMsmEngine> {
```

Comment:
It would be nice to have a small tests that shows:
- Building a `PlonkEngine` with `build_default`
- Building a `PlonkEngine` with `build`.

Reply:
added in https://github.com/privacy-scaling-explorations/halo2/pull/277/commits/6fa8b3a7c8eebe7ee29bac3c99b8d0d1e2c64f12